app/src/main/java/space/taran/arkrate/data/crypto/CryptoCurrencyRepo.kt

```diff
@@ -0,0 +1,51 @@
+package space.taran.arkrate.data.crypto
+
+import android.util.Log
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import space.taran.arkrate.data.CurrencyName
+import space.taran.arkrate.data.CurrencyRate
+import space.taran.arkrate.data.CurrencyRepo
+import java.util.concurrent.TimeUnit
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class CryptoCurrencyRepo @Inject constructor(
+    private val cryptoAPI: CryptoAPI
+) : CurrencyRepo {
+    private var cryptoWithRates: List<CurrencyRate>? = null
+        set(value) {
+            updatedTS = System.currentTimeMillis()
+            field = value
+        }
+    private var updatedTS: Long? = null
+
+    override suspend fun getCurrencyRate(): List<CurrencyRate> =
+        withContext(Dispatchers.IO) {
+            Log.d("T", "$dayInMillis")
+            if (cryptoWithRates == null ||
+                updatedTS!! + dayInMillis < System.currentTimeMillis()
+            ) {
+                cryptoWithRates = cryptoAPI.getCryptoRates().findUSDTPairs()
+            }
+
+            cryptoWithRates!!
+        }
+
+    override suspend fun getCurrencyName(): List<CurrencyName> = getCurrencyRate().map {
+        CurrencyName(it.code, name = "")
+    }
+
+    // api returns pairs like ETHBTC, ETHBNB, ETHTRX, ETHUSDT
+    // we only take USDT pairs
```

Comment:
Maybe better to store all pairs. This way, it would be possible to first look if we have direct `ETH->BTC` pair, and only fallback to computing rate via `ETH->USDT` then `USDT->BTC` if don't have direct pair.

Reply:
I can create separate issue for this, if necessary.

Reply:
Created https://github.com/ARK-Builders/ARK-Rate/issues/14